### DISCUSSION TITLE
Сustom PHP extensions & compiling and protecting PHP code

### DISCUSSION THREAD

**Discussion by JWprogrammer** [NativePHP/laravel#368] [Ideas]
Hi. Would it be possible to make a feature to add custom PHP extensions?
This would allow compiling PHP code into an extension to protect it from being copied and modified

**Comment by simonhamp** [comment]
In theory, if you can statically compile your extension, you can create a binary that includes it.

This isn't something I think we'll be adding to the NativePHP binaries any time soon though.

Bundling the source into an extension and then compiling that into the PHP binary is quite a complex approach that couples your code to the version of PHP you use.

This means you'd need to rebuild _each_ PHP binary for each platform that you want to support _every_ time you want to update your app.

This feels like overkill to me.

We've already got a working solution that will prevent your app's source code from being modified that doesn't introduce this level of complexity.

As for copying source code, my argument there is: your code - in whatever form - is on an untrusted client machine, so all bets are off. If someone _really_ wants to figure it out, they will.

If this is a deal breaker, I recommend you move those parts of the code to a server that you have greater control over.

**Reply by JWprogrammer** [reply]
Hi. Thank you for the response.
> We've already got a working solution that will prevent your app's source code from being modified that doesn't introduce this level of complexity.

I wonder how code modification protection would work?
Right now I can modify the application source code in PHP files after installation and the application works with the modifications.
Protection against this would be very good.


> Bundling the source into an extension and then compiling that into the PHP binary is quite a complex approach that couples your code to the version of PHP you use.

I have the code automatically compiled into an extension :)
And although the user's machine has access to the **compiled** app, the code cannot be copied like a ready-made **open source** project.

It would be nice if I could just add the .so/.dll extension to the ini parameters.
And, if NativePHP had something like middleware when building an application to run code compilation automatically, it would be very easy to build secure applications.

By the way, thank you for your hard work.
If we use NativePHP in our startup, we can support its development financially in the future
